Add Groovy layer

jdenen opened the pull request:
- Syntax highlighting
- REPL integration via `inferior-groovy-mode`
- Compiling/testing/task-running via `gradle-mode`

robbyoconnor commented:
Configure flycheck for groovy, there's already a checker for it.

jdenen commented:
@robbyoconnor Done. I'm not a flycheck user so thanks for pointing that out.

robbyoconnor commented:
> @robbyoconnor Done. I'm not a flycheck user so thanks for pointing that out.

Why are you not?!?!?! It's good!

d12frosted commented on layers/+lang/groovy/config.el:
```diff
@@ -0,0 +1,20 @@
+;;
+;;; License: GPLv3
+
+(defun groovy/gradle-clean-build ()
```
These functions should not be in `config.el` file. I think the best place is `funcs.el`([doc][1]). Also it would be good to 'guard' these functions by placing them inside of `(when (configuration-layer/package-usedp 'gradle-mode) FUNCS)`, so they don't error if someone calls them with `gradle-mode` excluded. 

  [1]: https://github.com/syl20bnr/spacemacs/blob/master/doc/LAYERS.org#funcsel

jdenen commented:
@d12frosted Done. Let me know if there's anything else I can do.

logc commented:
Is this ready to be merged? I just started to use Spacemacs a couple of months ago, and minutes ago discovered I would need a Groovy layer 😄

robbyoconnor commented:
@logc cherry-pick it into your local spacemacs ;)

logc commented:
@robbyoconnor thanks for the hint, didn't occur to me!

For my part, I'm happy to have this in a separate branch, but I'm also giving a :+1: to the "Done" comment, because, as a user, I think it is a useful layer.

robbyoconnor commented on layers/+lang/groovy/funcs.el:
```diff
@@ -0,0 +1,21 @@
+;;; funcs.el --- Groovy layer functions file for Spacemacs.
+;;
+;; Copyright (c) 2012-2016 Sylvain Benner & Contributors
```
Extend to 2017.

jdenen commented:
I changed a comment and one check failed. That seems indicative of a problem with the check but I don't know enough about it to be sure. I rebased onto develop and it continues to fail. Can anyone point me in a direction to get this check passing?

robbyoconnor commented:
It's not your fault. You can safely ignore it.

mxinden commented:
What is blocking this PR? Would love to see this merged into `develop`.

MerlinDMC commented:
@mxinden at least the renamed flycheck function in development :)

Fixed by:
```
diff --git a/layers/+lang/groovy/packages.el b/layers/+lang/groovy/packages.el
index 7debb168..0d589906 100644
--- a/layers/+lang/groovy/packages.el
+++ b/layers/+lang/groovy/packages.el
@@ -44,6 +44,6 @@
         "tt" 'gradle-single-test))))
 
 (defun groovy/post-init-flycheck ()
-  (spacemacs/add-flycheck-hook 'groovy-mode-hook))
+  (spacemacs/enable-flycheck 'groovy-mode-hook))
 
 ;;; packages.el ends here
```

madmod commented:
Please merge this!

robbyoconnor commented:
It has to be rebased first by @jdenen

jdenen commented:
Apologies, I didn't realize I was a blocker. I'll get this done shortly.

robbyoconnor commented:
I don't merge rights, but it'd be a requirement for me to recommend people cherry-pick this to their local environment :)

jdenen commented:
Closing since a496401e1f9452363152f5aa7c2858c7e7c91864 created a Groovy layer 🙌
